### ISSUE TITLE
AttributeError: 'HadoopFileSystem' object has no attribute 'close'

### ISSUE THREAD

**Issue by DavidKatz-il** [intake/filesystem_spec#429]
When running the following code:
```python
import fsspec
f = fsspec.open('hdfs:///path/to/file.csv').open()
...
```
I get the following error when the script exits (but complete successfully):
``` python
AttributeError: 'HadoopFileSystem' object has no attribute 'close'
Exception ignored in: 'pyarrow.lib.HadoopFileSystem.__dealloc__'
AttributeError: 'HadoopFileSystem' object has no attribute 'close'
```
I'm also getting the error when running:
```python
import fsspec
f = fsspec.open('hdfs:///path/to/file.csv').open()
...
f.close()
```
      
The versions in my environment are:
```
python: 3.7.8
fsspec: 0.8.0
pyarrow: 1.0.1
```

**Comment by martindurant** [comment]
Well I suppose that `close` can be added to the list of known attributes [here](https://github.com/intake/filesystem_spec/blob/master/fsspec/implementations/hdfs.py#L138); but interesting that this was not apparently needed before. Note that this is a method for the class instance, not the file, to close and free the connection, which is happening during garbage collection.

**Comment by DavidKatz-il** [comment]
I get this error when working on [this](https://github.com/pandas-dev/pandas/issues/18199).

**Comment by martindurant** [comment]
Would you like to provide the PR?
The HDFS interface isn't even tested, because it would require to be run *within* the hadoop cluster (as opposed to webHDFS). I guess it could be done with the right docker invocation?

**Comment by DavidKatz-il** [comment]
I will be happy to make a PR.
Are you familiar with some docker file that will give me the option to use `pyarrow` from the local machine for reading / writing within the hdfs container?

**Comment by martindurant** [comment]
Presumably you would need to mount the local directory into the countainer?

**Comment by DavidKatz-il** [comment]
Hi @martindurant, 
Adding 'close' to the list of known attributes didn't solve the issue.
The last item that is called in `__getattribute__`  before exit is `async_impl`, the function return `super().__getattribute__(item)`.
After that, we get the `AttributeError`.

**Comment by martindurant** [comment]
Can add that one too :)
I might at some point come up with a general and foolproof way of making these "pass-though" classes, but apparently not yet.

**Comment by DavidKatz-il** [comment]
When adding `async_impl` i get the following error:
``` python
  File "/.../filesystem_spec/fsspec/registry.py", line 227, in filesystem
    return cls(**storage_options)
  File "/.../filesystem_spec/fsspec/spec.py", line 63, in __call__
    if obj.async_impl:
  File "/.../filesystem_spec/fsspec/implementations/hdfs.py", line 175, in __getattribute__
    return getattr(pahdfs, item)
AttributeError: 'HadoopFileSystem' object has no attribute 'async_impl'
```

**Comment by martindurant** [comment]
Yep, looks like I was wrong there... I'll have to experiment with it myself when I can get the chance, unless you want to poke around. Did you work out a reasonable testing hook without having a read hadoop?

**Comment by DavidKatz-il** [comment]
Im running in a [dask docker](https://github.com/dask/dask/tree/master/continuous_integration/hdfs).
I also write some tests (base on dask-hdfs-tests):
``` python
import os
import posixpath

import pytest
import fsspec

try:
    import pyarrow
except ImportError:
    pyarrow = None


basedir = "/tmp/test-fsspec"
data = b"\n".join([b"some test data"] * 1000)


@pytest.fixture
def hdfs(request):
    hdfs = pyarrow.hdfs.connect(host="localhost", port=8020)

    if hdfs.exists(basedir):
        hdfs.rm(basedir, recursive=True)
    
    hdfs.mkdir(basedir)
    
    with hdfs.open(basedir + "/file", 'wb') as f:
        f.write(data)
    
    yield hdfs

    if hdfs.exists(basedir):
        hdfs.rm(basedir, recursive=True)


# This mark doesn't check the minimum pyarrow version.
require_pyarrow = pytest.mark.skipif(not pyarrow, reason="pyarrow not installed")


def test_ls(hdfs):
    h = fsspec.filesystem("hdfs")
    out = [f['name'] for f in h.ls(basedir)]
    assert out == hdfs.ls(basedir)


def test_walk(hdfs):
    h = fsspec.filesystem("hdfs")
    out = h.walk(basedir)
    assert list(out) == list(hdfs.walk(basedir))


def test_isdir(hdfs):
    h = fsspec.filesystem("hdfs")
    assert h.isdir(basedir)
    assert not h.isdir(basedir + "/file")


def test_exists(hdfs):
    h = fsspec.filesystem("hdfs")
    assert not h.exists(basedir + "/notafile")


def test_read(hdfs):
    h = fsspec.filesystem("hdfs")
    out = basedir + "/file"
    with h.open(out, "rb") as f:
        assert f.read() == data
    with h.open(out, "rb", block_size=0) as f:
        assert f.read() == data
    with h.open(out, "rb") as f:
        assert f.read(100) + f.read() == data
```
Those tests will have to execute within a docker.

**Comment by martindurant** [comment]
Those tests do all pass for me - was one of them failing, or was your `open` from other work?

**Comment by DavidKatz-il** [comment]
Those tests are for `fsspec` with `hdfs`, they all passed.
I don't know how to make a test for errors after exit.

When running:

``` python
import fsspec
h = fsspec.filesystem("hdfs")
```

The error pop's after exit.

**Comment by martindurant** [comment]
I haven't been able to make the error happen, with explicit garbage collection or normal shutdown.

We could add
```python
    def close(self):
        self.pahdfs.close()
```
(and add `"close"` to the first list in `__getattribute__`).
I have a shonky way to run the tests against docker hadoop, may spend a little time making it better, if I get the chance.

**Comment by DavidKatz-il** [comment]
I found an old [issue](https://github.com/intake/filesystem_spec/issues/111) that is still open with the same error.

> I haven't been able to make the error happen, with explicit garbage collection or normal shutdown.
> 
> We could add
> 
> ```python
>     def close(self):
>         self.pahdfs.close()
> ```
> 
> (and add `"close"` to the first list in `__getattribute__`).    

This didn't solve it.

when running :
```python
import fsspec
h = fsspec.filesystem('hdfs')
...
del h
```
We are still seeing the error.

When adding "close" to the list of known attributes and calling "close" before exit:
```python 
import fsspec
h = fsspec.filesystem('hdfs')
...
h.close()
```
The error didn't show up.

**Comment by martindurant** [comment]
> The error didn't show up.

So this is the solution? Happy to include it.

**Comment by DavidKatz-il** [comment]
> > The error didn't show up.
> 
> So this is the solution? Happy to include it.

This is a solution solves (must manually calling "close"):
```python
import fsspec
h = fsspec.filesystem('hdfs')
...
h.close()
```
But it's not solving (even when calling "close"):
``` python
import fsspec
f = fsspec.open('hdfs:///path/to/file.csv').open()
...
f.close()
```
Also when reading `hdfs` with `pandas` we still getting the error.

**Comment by martindurant** [comment]
Sorry, I'm confused now! Is it close on the file or the filesystem that's causing the problem? Perhaps the linked PR solves this?
Note that 
```
f = fsspec.open('hdfs:///path/to/file.csv').open()
...
f.close()
```
is better as
```
with fsspec.open('hdfs:///path/to/file.csv') as f
    f...
```
unless you have a good reason to handle the close yourself.

**Comment by martindurant** [comment]
My PR did this slightly differently, perhaps it works better?

It may be that this is an interpreter teardown issue, in which case we should make a `weakref.finalize` in the `__init__` method.

**Comment by DavidKatz-il** [comment]
> My PR did this slightly differently, perhaps it works better?

The behavior of your PR is the same, it's doing the same thing differently.

**Comment by martindurant** [comment]
How about with https://github.com/intake/filesystem_spec/pull/435/commits/289b77ca92443273ca83b1e368c34ae1067578cc ?

**Comment by DavidKatz-il** [comment]
> How about with [289b77c](https://github.com/intake/filesystem_spec/commit/289b77ca92443273ca83b1e368c34ae1067578cc) ?

Solved...

**Comment by martindurant** [comment]
! :)

**Comment by DavidKatz-il** [comment]
You can also close #111.
